### PR TITLE
Add forceNestedToLoose option to lint-list-item-spacing plugin

### PR DESCRIPTION
**Pull request by barraq**
# Description

Introduces an option to disable nested list to be loose when being tight list.

# Example

When this rule is turned on, the following file
`invalid.md` is **not** ok:

```markdown
A tight list:

-   Wrapped
    item
-   item 2
-   item 3

A loose list:

-   item 1

-   item 2

-   item 3

A nested list:

-   item 1
    - item 1.1
    - item 1.2
-   item 2
-   item 3
```

```text
4:9-5:1: Missing new line after list item
5:11-6:1: Missing new line after list item
11:1-12:1: Extraneous new line after list item
13:1-14:1: Extraneous new line after list item
20:15-21:1: Missing new line after list item
21:11-22:1: Missing new line after list item
```

When this rule is `{ forceNestedToLoose: false }`, the following file
`valid.md` is ok:

```markdown
A tight list:

-   item 1
-   item 2
-   item 3

A loose list:

-   Wrapped
    item

-   item 2

-   item 3

A nested list:

-   item 1
    - item 1.1
    - item 1.2
-   item 2
-   item 3
```

### CODE REVIEW

File: packages/remark-lint-list-item-spacing/index.js [hunk 1]
```diff
@@ -86,9 +143,20 @@ function listItemSpacing(ast, file) {
     items.forEach(warn);
 
     function infer(item) {
+      console.log(item.children);
+
```
Comment (wooorm) [inline packages/remark-lint-list-item-spacing/index.js]: Drop the `console.log`!

File: packages/remark-lint-list-item-spacing/index.js [hunk 2]
```diff
@@ -7,11 +7,16 @@
  *   Warn when list looseness is incorrect, such as being tight
  *   when it should be loose, and vice versa.
  *
+ *   Options: `Object`.
```
Comment (wooorm) [inline packages/remark-lint-list-item-spacing/index.js]: Move this to just above the definition of the actual options.
Could you also mention it’s optional? Like so: ``Options: optional `Object`.``

File: packages/remark-lint-list-item-spacing/index.js (hunk 1 above)
Comment (barraq) [inline packages/remark-lint-list-item-spacing/index.js]: ho my bad! forgot this one 🙈

**Comment by barraq** [comment]
> Why is this rule about one paragraph and one list? Why not about:

* `a list-item with a paragraph and two lists?` because then it will be really a multi-line list item since to have 2 two lists you will need a paragraph (or something else) in between.

* a list with a paragraph and other content, like a block quote? because then it will be really a multi-line item (not a single line one).

* all the possible paragraphs in a list-item? You mean multi-line paragraph? cause then again it will be a multi-line list item which fall in the category of loose list.

Does it make sense?

Basically to me, the list below is tight and not loose because each item is a single line but one item has a nested list.

```markdown
-   item 1
    - item 1.1
    - item 1.2
-   item 2
-   item 3
```

**Comment by wooorm** [comment]
How about:

```markdown
-   item 1
    > blocquote
-   item 2
-   item 3
```

...or:

```markdown
-   item a
    - list
    continued
-   item 2
-   item 3
```

...or:

```markdown
-   item a
    > blocquote
    item b
-   item 2
-   item 3
```

...or:

```markdown
-   item a
    ```foo
    bar();
    ```
-   item 2
-   item 3
```

**Comment by barraq** [comment]
All the below should not be affected by this option (i.e.  `forceNestedToLoose`) since they are not nested lists; also it is a loose list so it should raise a warning.

```
-   item 1
    > blocquote
-   item 2
-   item 3
```

```
-   item a
    ```foo
    bar();
    ```
-   item 2
-   item 3
```

```
-   item a
    > blocquote
    item b
-   item 2
-   item 3
```

The case below is interesting, with the option set to `false` it will not complain but by default it will raise:

```
-   item a
    - list
    continued
-   item 2
-   item 3
```

```shell
  3:14-4:1: Missing new line after list item  list-item-spacing  remark-lint
  4:11-5:1: Missing new line after list item  list-item-spacing  remark-lint
```

**Comment by barraq** [comment]
Maybe the option should be named differently (not sure about the name though).  Basically, the idea is to allow only the nested-single-line items to be tight, so the construct:

```
- some single line content
  - some single line content
  - some single line content
- some single line content
- some single line content
```

This can be helpful when creating table of content, or basic enumeration. As soon as one list item is multi line then the plugin should apply as it is.

**Comment by barraq** [comment]
@wooorm do you want to proceed differently?

**Comment by wooorm** [comment]
Sorry, busy with work!

I first thought you wanted a rule that essentially checked for black lines, instead of the current behaviour, but I now see you just want to allow lists with a paragraph “header”.

The [`no-missing-blank-lines`](https://github.com/freeletics/remark-lint/tree/master/packages/remark-lint-no-missing-blank-lines) rule has a similar option: `exceptTightLists`, which allows “missing” blank lines. It basically allows:

```markdown
- Paragraph
  *  List.
```

Shouldn’t we just add something like that here too? It’s not exactly what you want to do, but it will allow style and is in line with another rule?

**Comment by barraq** [comment]
It would be good to find a name for that option and use it there too indeed.
Should we name it: `allowListWithParagraphHeader`?
If you are fine with the naming then I'll rebase and also add another commit for [no-missing-blank-lines](https://github.com/freeletics/remark-lint/tree/master/packages/remark-lint-no-missing-blank-lines) plugin.

**Comment by wooorm** [comment]
Hi @barraq, sorry that this took a while!

I spent some more time thinking on this, and I think I have just pushed a solution that works for you.
See https://github.com/wooorm/remark-lint/commit/95361061cef0db1dc47535ce8e22e0cee44e5b4f for more info.

Let me know if it works!

**Comment by wooorm** [comment]
Published as [1.1.0](https://www.npmjs.com/package/remark-lint-list-item-spacing)!
